### PR TITLE
Quincy: osd/scrub: verify SnapMapper consistency

### PR DESCRIPTION
**Pull request by ronen-fr**
Whenever the scrubber accesses the SnapMapper for the snaps of a specific
clone, the mapper will now verify that the snaps have the required
mapping DB entries (the 'SNA_' keys).

A backport of https://github.com/ceph/ceph/pull/47388
Backport tracker: https://tracker.ceph.com/issues/59537

Note: contains cherry-picked fixes to compilation issues
(from PRs by Adam Emerson & Kefu)

Signed-off-by: Ronen Friedman [rfriedma@redhat.com](mailto:rfriedma@redhat.com)

### CODE REVIEW

**Comment by ronen-fr** [comment]
Note: I have verified that the 6 compilation-fixes commits (cherry-picked) were required not only in my
build environments.

**Comment by ronen-fr** [comment]
Multiple Teuthology runs - OK.

**Comment by athanatos** [comment]
Why do we need the xlist iterator backports?

**Comment by athanatos** [comment]
I don't see an explanation of *why* these changes require the apparently unrelated std::iterator changes or the LogEntry, Journald, and neorados changes to be backported.  @ronen-fr I particularly need to know exactly why librados.hpp needs to be modified as part of this PR -- that's not something I like to do casually.  Are these backports needed for current quincy without the scrub commits to build on some platforms?  If not, what about the scrub commits makes this necessary?

**Comment by ronen-fr** [comment]
> "including compilation fixes that were required, but are not relevant to the snap-mapper issue" <-- From this I suppose that quincy is already broken? On which targets?

@athanatos - yes - Quincy "as is" won't compile on Fedora 37 & 38 (g++)

**Review by athanatos — APPROVED** [review]
Ok, the changes to fix compile issues seem ok to me, and I don't think the librados.hpp change will be a problem (@rzarzynski can you double check me on that one?).

The scrub changes also seem to be correct as far as I can tell, but I'd like someone else to review them as well.  Unfortunately, they really do cross enough refactors that we probably can't get a cleaner backport.

- SnapMapper.h/cc seem to match main except for changes added for crimson
- osd_types_fmt.h mostly matches main except for the bits I pointed out
- The changes to pg_scrubber.cc are specific to this backport, but they look correct to me.
- The PR includes relevant tests, so the test results can probably be relied upon to avoid regressions

I'd be more comfortable if you create a quick PR for main with the osd_types_fmt.h changes not yet in main so we can merge that first and then cherry-pick the commit -- I'd prefer that quincy not have build fixes not in main.

Larger question not related to this PR: why do libfmt version bumps keep breaking our usage of the library?  Is the api simply that unstable?

**Comment by ronen-fr** [comment]
@athanatos :
> I'd be more comfortable if you create a quick PR for main with the osd_types_fmt.h changes not yet in main so we can merge that first and then cherry-pick the commit -- I'd prefer that quincy not have build fixes not in main.

Agree - created https://github.com/ceph/ceph/pull/52337

**Comment by ronen-fr** [comment]
> Larger question not related to this PR: why do platform libfmt version bumps keep breaking our usage of the library? Is the api simply that unstable?

@tchaikov, who handled many of these changes, can probably give a more educated answer. I would suggest that part of the problem was that we were late to move to the stable versions of the library (mostly as a result of Centos 8 issues).

File: src/osd/osd_types_fmt.h [hunk 1]
```diff
@@ -110,8 +114,239 @@ struct fmt::formatter<pg_t> {
...
+      fmt::format_to(ctx.out(), "\n");
+    }
+    return fmt::format_to(ctx.out(), "}}");
+  }
+
+  bool debug_log{false};
+};
+
+template <>
+struct fmt::formatter<scrub_level_t> : fmt::formatter<std::string_view> {
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:332]: agree. Will be removed.

File: src/osd/osd_types_fmt.h [hunk 2]
```diff
@@ -110,8 +114,239 @@ struct fmt::formatter<pg_t> {
...
+    }
+  }
+};
+
+template <>
+struct fmt::formatter<pg_info_t> {
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+
+  template <typename FormatContext>
+  auto format(const pg_info_t& pgi, FormatContext& ctx) const
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:172]: will be handled, as requested, by backporting https://github.com/ceph/ceph/pull/52337 (once approved)

File: src/osd/osd_types_fmt.h [hunk 3]
```diff
@@ -110,8 +114,239 @@ struct fmt::formatter<pg_t> {
...
+  bool verbose{false};
+};
+
+template <>
+struct fmt::formatter<ScrubMap::object> {
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+
+  ///\todo: consider passing the 'D" flag to control snapset dump
+  template <typename FormatContext>
+  auto format(const ScrubMap::object& so, FormatContext& ctx) const
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:268]: same here

File: src/osd/osd_types_fmt.h [hunk 4]
```diff
@@ -110,8 +114,239 @@ struct fmt::formatter<pg_t> {
...
+    auto it = ctx.begin();
+    if (it != ctx.end() && *it == 'D') {
+      debug_log = true;	 // list the objects
+      ++it;
+    }
+    return it;
+  }
+
+  template <typename FormatContext>
+  auto format(const ScrubMap& smap, FormatContext& ctx) const
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:311]: will be modified

File: src/osd/osd_types_fmt.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:172]: done

File: src/osd/osd_types_fmt.h [hunk 5]
```diff
@@ -110,8 +114,239 @@ struct fmt::formatter<pg_t> {
...
+    auto it = ctx.begin();
+    if (it != ctx.end() && *it == 'D') {
+      verbose = true;
+      ++it;
+    }
+    return it;
+  }
+
+  template <typename FormatContext>
+  auto format(const SnapSet& snps, FormatContext& ctx) const
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:214]: done

File: src/osd/osd_types_fmt.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h:268]: done (now part of a c-p)

**Review by rzarzynski — APPROVED** [review]

**Comment by Matan-B** [comment]
@ronen-fr
IIUC, `ceph_test_snap_mapper` is not actually being run. Shouldn't we add a new task in the rados suite to include the test that verifies the changes here?

**Comment by ronen-fr** [comment]
> IIUC, `ceph_test_snap_mapper` is not actually being run. Shouldn't we add a new task in the rados suite to include the test that verifies the changes here?

@Matan-B : I missed that fact... Yes - we should. I'll contact you offline.

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/QUINCY#httpstrellocomcw1wxAcJO1814-wip-yuri8-testing-2023-07-24-0819-quincy

**Pull request merged**